Introduce ProcessSyncClient, and move broadcastMainFrameURLChangeToOtherProcesses to it

beidson opened the pull request:
#### 1b6a2d49b73b499f03bfd3270d7d55d6e78999cc
<pre>
Introduce ProcessSyncClient, and move broadcastMainFrameURLChangeToOtherProcesses to it
<a href="https://rdar.apple.com/139370028">rdar://139370028</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282700">https://bugs.webkit.org/show_bug.cgi?id=282700</a>

Reviewed by Alex Christensen.

With site isolation, there will be many fields of data set by one process that need to be broadcast to all processes.

The existing, ad-hoc one-off &quot;mainFrameURLChange&quot; was the first of these. It won&apos;t be the last.

We need a better mechanism to handle synchronizing all this data.
And that mechanism is the ProcessSyncClient that is added to Page.

For now, I move the main frame URL broadcast away from FrameLoaderClient to ProcessSyncClient.
In the next patch, I autogenerate the relevant code.
Then after that, I&apos;ll start adding more properties that need to be broadcast.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
(WebCore::EmptyFrameLoaderClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebCore/loader/EmptyFrameLoaderClient.h:

* Source/WebCore/loader/LocalFrameLoaderClient.h:

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::documentURLDidChange):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::processSyncClient):
(WebCore::Page::processSyncClient const):

* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:

* Source/WebCore/page/ProcessSyncClient.h: Added.
(WebCore::ProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses):

* Source/WebKit/Sources.txt:

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:

* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.cpp: Added.
(WebKit::WebProcessSyncClient::WebProcessSyncClient):
(WebKit::WebProcessSyncClient::protectedPage const):
(WebKit::WebProcessSyncClient::siteIsolationEnabled):
(WebKit::WebProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses):
* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.h: Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
(WebFrameLoaderClient::updateOpener):

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: <a href="https://commits.webkit.org/286247@main">https://commits.webkit.org/286247@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02c9ac1dc3e763e5af20069e331b9857bc8650b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75314 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/54762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28154 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77431 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2538 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/79777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78381 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49301 "") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/64706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/46697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22199 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67708 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81261 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2644 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2795 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64702 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/66647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16578 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/10596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2605 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2630 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/ProcessSyncClient.h:
```diff
@@ -0,0 +1,49 @@
+class URL;
+}
+
+using WTF::URL;
```
Seems like wtf/Forward.h would be better than all this.

achristensen07 commented on Source/WebCore/page/ProcessSyncClient.h:
```diff
@@ -0,0 +1,49 @@
+    ProcessSyncClient() = default;
+    virtual ~ProcessSyncClient() = default;
+
+    virtual void broadcastMainFrameURLChangeToOtherProcesses(const URL&) { }
```
= 0 and let's make a EmptyProcessSyncClient

beidson replied:
Philosophical disagreement here. 
This is nothing more than WebCore -> WebKit2 bridge, not a more traditional platform client.
Why rope ourselves into having to change 2+ code sites every method added instead of just the one that matters?

beidson committed:
Introduce ProcessSyncClient, and move broadcastMainFrameURLChangeToOtherProcesses to it
rdar://139370028
https://bugs.webkit.org/show_bug.cgi?id=282700

Reviewed by Alex Christensen.

With site isolation, there will be many fields of data set by one process that need to be broadcast to all processes.

The existing, ad-hoc one-off "mainFrameURLChange" was the first of these. It won't be the last.

We need a better mechanism to handle synchronizing all this data.
And that mechanism is the ProcessSyncClient that is added to Page.

For now, I move the main frame URL broadcast away from FrameLoaderClient to ProcessSyncClient.
In the next patch, I autogenerate the relevant code.
Then after that, I'll start adding more properties that need to be broadcast.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
(WebCore::EmptyFrameLoaderClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebCore/loader/EmptyFrameLoaderClient.h:

* Source/WebCore/loader/LocalFrameLoaderClient.h:

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::documentURLDidChange):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::processSyncClient):
(WebCore::Page::processSyncClient const):

* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:

* Source/WebCore/page/ProcessSyncClient.h: Added.
(WebCore::ProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses):

* Source/WebKit/Sources.txt:

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::broadcastMainFrameURLChangeToOtherProcesses): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:

* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.cpp: Added.
(WebKit::WebProcessSyncClient::WebProcessSyncClient):
(WebKit::WebProcessSyncClient::protectedPage const):
(WebKit::WebProcessSyncClient::siteIsolationEnabled):
(WebKit::WebProcessSyncClient::broadcastMainFrameURLChangeToOtherProcesses):
* Source/WebKit/WebProcess/WebCoreSupport/WebProcessSyncClient.h: Added.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
(WebFrameLoaderClient::updateOpener):

* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: https://commits.webkit.org/286247@main
